### ISSUE TITLE
Bad type inference leading to crash at runtime on scala3

### ISSUE THREAD

**Issue by OlegYch** [slick/slick#2674]
**DESCRIPTION:**

Using slick 3.5.0-M2 (built for scala2.13) on scala 3.0.0 - 3.3.0-RC3
https://scastie.scala-lang.org/rQ5sOqVyTXmn7BuqMt3hkw
I get ClassCastException at runtime

**STEPS TO REPRODUCE:**

```scala
import slick.memory.MemoryProfile.api._
@main def main = 
	val xx = Rep.None[Int].getOrElse(0)
```

**ACTUAL RESULT:**

`java.lang.ClassCastException: class slick.lifted.Rep$$anon$1 cannot be cast to class slick.lifted.ConstColumn `

reported at https://github.com/lampepfl/dotty/issues/17074 as well, as i'm not sure where should it be fixed

**Comment by nafg** [comment]
What is the objective of the snippet?

**Comment by OlegYch** [comment]
@nafg no particular objective, just a minimized sample

**Comment by nafg** [comment]
Can you try it on the `nafg/dottyquery` branch?

**Comment by OlegYch** [comment]
@nafg it doesn't seem to compile
```
sbt:slick-root> slick/console
[info] running fmpp.tools.CommandLine --expert -q -S D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala -O D:\Distrib\Coding\Java\scala\slick\slick\target\scala-3.1.2\src_managed\main "--replace-extensions=fm, scala" -M "execute(**/*.fm), ignore(**/*)"
[info] compiling 175 Scala sources to D:\Distrib\Coding\Java\scala\slick\slick\target\scala-3.1.2\classes ...
[warn] -- Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala-3\slick\lifted\ShapedValue.scala:93:53 -------------
[warn] 93 |  implicit def someToTuple[T]: ToTuple[Some[T], T] = identity
[warn]    |                                                     ^^^^^^^^
[warn]    |method identity is eta-expanded even though slick.lifted.ToTuple[Some[T], T] does not have the @FunctionalInterface annotation.
[warn] -- Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala-3\slick\lifted\ShapedValue.scala:97:57 -------------
[warn] 97 |  implicit def optionToTuple[T]: ToTuple[Option[T], T] = identity
[warn]    |                                                         ^^^^^^^^
[warn]    |method identity is eta-expanded even though slick.lifted.ToTuple[Option[T], T] does not have the @FunctionalInterface annotation.
[error] -- Error: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala-3\slick\lifted\TableQuery.scala:53:53 ----------------
[error] 53 |    '{ TableQuery.apply[E](${ cons.asExprOf[Tag => E] }) }
[error]    |                                                     ^
[error]    |no implicit argument of type quoted.Type[slick.lifted.Tag => E] was found for parameter x$2 of method asExprOf in trait TreeMethods
[error]    |
[error]    |where:    E is a type in method applyExpr with bounds <: slick.lifted.AbstractTable[?]
[error] -- [E007] Type Mismatch Error: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\ast\Node.scala:75:4 -------
[error] 75 |    if(seenType || _type != UnassignedType) buildCopy else this
[error]    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |    Found:    slick.ast.Node
[error]    |    Required: Node.this.Self
[error] -- [E007] Type Mismatch Error: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\ast\Node.scala:79:18 ------
[error] 79 |    val n: Self = if(seenType && newType != _type) buildCopy else this
[error]    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |                  Found:    slick.ast.Node
[error]    |                  Required: Node.this.Self
[error] -- Error: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\ast\Type.scala:189:56 --------------------------
[error] 189 |  private[this] val arrayClassTag = mkClassTag[Array[?]]
[error]     |                                                        ^
[error]     |                                                        No ClassTag available for Array[?]
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\HsqldbProfile.scala:44:4 ------
[warn] 44 |    - SqlCapabilities.sequenceCurr
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\HsqldbProfile.scala:45:4 ------
[warn] 45 |    - JdbcCapabilities.insertOrUpdate
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\JdbcModelBuilder.scala:278:43 -----------
[warn] 278 |        dbType.map(SqlProfile.ColumnOption.SqlType) ++
[warn]     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[warn]     |The method `apply` is inserted. The auto insertion will be deprecated, please write `slick.sql.SqlProfile.ColumnOption.SqlType.apply` explicitly.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:66:4 ------
[warn] 66 |    - RelationalCapabilities.foreignKeyActions
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:67:4 ------
[warn] 67 |    - JdbcCapabilities.insertOrUpdate
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:68:4 ------
[warn] 68 |    - JdbcCapabilities.booleanMetaData
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:69:4 ------
[warn] 69 |    - JdbcCapabilities.distinguishesIntTypes
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:70:4 ------
[warn] 70 |    - JdbcCapabilities.supportsByte
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:71:4 ------
[warn] 71 |    - JdbcCapabilities.returnMultipleInsertKey
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:72:4 ------
[warn] 72 |    - JdbcCapabilities.insertMultipleRowsWithSingleStatement
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:114:6 -----
[warn] 114 |      - Phase.fixRowNumberOrdering
[warn]     |      ^
[warn]     |      Rest of line starts with an operator;
[warn]     |      it is now treated as a continuation of the previous expression in parentheses,
[warn]     |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\OracleProfile.scala:115:6 -----
[warn] 115 |      + Phase.rewriteBooleans + new RemoveSubqueryOrdering
[warn]     |      ^
[warn]     |      Rest of line starts with an operator;
[warn]     |      it is now treated as a continuation of the previous expression in parentheses,
[warn]     |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\PostgresProfile.scala:56:4 ----
[warn] 56 |    - JdbcCapabilities.insertOrUpdate
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\PostgresProfile.scala:57:4 ----
[warn] 57 |    - JdbcCapabilities.insertOrUpdateWithPrimaryKeyOnly
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\PostgresProfile.scala:58:4 ----
[warn] 58 |    - JdbcCapabilities.nullableNoDefault
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\PostgresProfile.scala:59:4 ----
[warn] 59 |    - JdbcCapabilities.supportsByte
[warn]    |    ^
[warn]    |    Rest of line starts with an operator;
[warn]    |    it is now treated as a continuation of the previous expression in parentheses,
[warn]    |    not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:74:6 ---
[warn] 74 |      - JdbcCapabilities.forceInsert
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:75:6 ---
[warn] 75 |      - JdbcCapabilities.returnInsertOther
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:76:6 ---
[warn] 76 |      - JdbcCapabilities.insertOrUpdate
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:77:6 ---
[warn] 77 |      - SqlCapabilities.sequence
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:78:6 ---
[warn] 78 |      - JdbcCapabilities.supportsByte
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:79:6 ---
[warn] 79 |      - JdbcCapabilities.returnMultipleInsertKey
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[warn] -- Migration Warning: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\jdbc\SQLServerProfile.scala:80:6 ---
[warn] 80 |      - JdbcCapabilities.insertMultipleRowsWithSingleStatement
[warn]    |      ^
[warn]    |      Rest of line starts with an operator;
[warn]    |      it is now treated as a continuation of the previous expression in parentheses,
[warn]    |      not as a separate statement.
[error] -- [E007] Type Mismatch Error: D:\Distrib\Coding\Java\scala\slick\slick\src\main\scala\slick\lifted\Shape.scala:280:5 --
[error] 280 |    }
[error]     |     ^
[error]     |Found:    slick.lifted.CaseClassShape[P, LiftedTuple, P, PlainTuple, PlainCaseClass] {...}
[error]     |Required: slick.lifted.CaseClassShape[P, LiftedTuple, LiftedCaseClass, PlainTuple, PlainCaseClass]
[error]     |
[error]     |where:    LiftedCaseClass is a type in class CaseClassShape with bounds <: P
[error]     |          P               is a type in class CaseClassShape with bounds >: PlainCaseClass | LiftedCaseClass and <: Product
[error]     |          PlainCaseClass  is a type in class CaseClassShape with bounds <: P
[warn] 25 warnings found
[error] 5 errors found
[error] (slick / Compile / compileIncremental) Compilation failed
[error] Total time: 15 s, completed Mar 10, 2023 12:21:49 AM
sbt:slick-root>
```

**Comment by nafg** [comment]
You can stick it somewhere in the source code and see what type it infers, maybe with Metals?

Another trick is you do something like

```scala
val x = ...

x: Int
```

and you see from the type mismatch error what it inferred `x` to be.


That said, if you can get those errors to compile on scala 2 & 3 it would be appreciated :D

**Comment by OlegYch** [comment]
the code requires `Shape`, so at least that should compile...

**Comment by OlegYch** [comment]
fixed on 3.5.0-M4
